File: server/src/metrics.js
export function readCgroupCpu(read) {
  const stat = read("/sys/fs/cgroup/cpu.stat");
  const m = /usage_usec\s+(\d+)/.exec(stat);
  return { usageUsec: m ? Number(m[1]) : 0 };
}

export function readCgroupMemory(read) {
  const current = Number((read("/sys/fs/cgroup/memory.current") || "0").trim());
  const maxRaw = (read("/sys/fs/cgroup/memory.max") || "max").trim();
  return { current, max: maxRaw === "max" ? Infinity : Number(maxRaw) };
}

export function readNetDev(read) {
  const text = read("/proc/net/dev");
  let rxBytes = 0, txBytes = 0;
  for (const line of text.split("\n").slice(2)) {
    const parts = line.trim().split(":");
    if (parts.length !== 2) continue;
    const iface = parts[0].trim();
    if (iface === "lo") continue;
    const nums = parts[1].trim().split(/\s+/).map(Number);
    rxBytes += nums[0] || 0;
    txBytes += nums[8] || 0;
  }
  return { rxBytes, txBytes };
}

export function computeCpuPercent(prev, cur, elapsedMs, numCpus = 1) {
  if (!elapsedMs) return 0;
  const deltaUsec = (cur.usageUsec - prev.usageUsec);
  if (deltaUsec <= 0) return 0;
  const cpuSec = deltaUsec / 1e6;
  const wallSec = elapsedMs / 1000;
  return (cpuSec / wallSec / Math.max(1, numCpus)) * 100;
}

File: backend/internal/metrics/metrics.go
package metrics

import (
	"os"
	"regexp"
	"strconv"
	"strings"
)

type Reader = func(string) string

func readFile(path string) string {
	b, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return string(b)
}

func ReadFileFn() Reader { return readFile }

var cpuUsageRe = regexp.MustCompile(`usage_usec\s+(\d+)`)

func ReadCgroupCPU(read Reader) uint64 {
	m := cpuUsageRe.FindStringSubmatch(read("/sys/fs/cgroup/cpu.stat"))
	if len(m) < 2 {
		return 0
	}
	n, _ := strconv.ParseUint(m[1], 10, 64)
	return n
}

func ReadCgroupMemory(read Reader) (current, max uint64, maxSet bool) {
	c, _ := strconv.ParseUint(strings.TrimSpace(read("/sys/fs/cgroup/memory.current")), 10, 64)
	mr := strings.TrimSpace(read("/sys/fs/cgroup/memory.max"))
	if mr == "max" || mr == "" {
		return c, 0, false
	}
	mx, _ := strconv.ParseUint(mr, 10, 64)
	return c, mx, true
}

func ReadNetDev(read Reader) (rx, tx uint64) {
	text := read("/proc/net/dev")
	lines := strings.Split(text, "\n")
	if len(lines) < 3 {
		return 0, 0
	}
	for _, line := range lines[2:] {
		parts := strings.SplitN(strings.TrimSpace(line), ":", 2)
		if len(parts) != 2 {
			continue
		}
		iface := strings.TrimSpace(parts[0])
		if iface == "lo" {
			continue
		}
		nums := strings.Fields(parts[1])
		if len(nums) < 16 {
			continue
		}
		r, _ := strconv.ParseUint(nums[0], 10, 64)
		t, _ := strconv.ParseUint(nums[8], 10, 64)
		rx += r
		tx += t
	}
	return rx, tx
}
